drivers/gpio/gpio_tca9538.c

```diff
@@ -0,0 +1,541 @@
+	 */
+	if ((flags & (GPIO_DS_ALT_LOW | GPIO_DS_ALT_HIGH)) != 0) {
+		return -ENOTSUP;
+	}
+
+	/* Open drain and open source not supported */
+	if ((flags & GPIO_OPEN_DRAIN) != 0 || (flags & GPIO_OPEN_SOURCE) != 0) {
+		return -ENOTSUP;
+	}
+
+	/* The TCA9538 has no internal pull up support */
+	if (((flags & GPIO_PULL_UP) != 0) || ((flags & GPIO_PULL_DOWN) != 0)) {
+		return -ENOTSUP;
+	}
+
+	/* The TCA9538 does not support internal debounce */
+	if ((flags & GPIO_INT_DEBOUNCE) != 0) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/* Ensure either Output or Input is specified */
+	if ((flags & GPIO_OUTPUT) != 0) {
+		pins->dir &= ~BIT(pin);
+		if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+			pins->output &= ~BIT(pin);
+			data_first = true;
+		} else if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			pins->output |= BIT(pin);
+			data_first = true;
+		}
+	} else if ((flags & GPIO_INPUT) != 0) {
+		pins->dir |= BIT(pin);
+	} else {
+		rc = -ENOTSUP;
+		goto out;
+	}
+
+	/* Set output values */
+	if (data_first) {
+		tx_dat[0] = TCA9538_OUTPUT_PORT;
+		tx_dat[1] = pins->output;
+		rc = i2c_write(cfg->i2c_dev, tx_dat, sizeof(tx_dat),
+			       cfg->i2c_addr);
+	}
+
+	if (rc == 0) {
+		/* Set pin directions */
+		tx_dat[0] = TCA9538_CONFIGURATION;
+		tx_dat[1] = pins->dir;
+		rc = i2c_write(cfg->i2c_dev, tx_dat, sizeof(tx_dat),
+			       cfg->i2c_addr);
+	}
+out:
+	k_sem_give(&drv_data->lock);
+	return rc;
+}
+
+static int gpio_tca9538_port_get(const struct device *dev,
+				 gpio_port_value_t *value)
+{
+	const struct tca9538_config *cfg = dev->config;
+	struct tca9538_drv_data *drv_data = dev->data;
+	uint8_t input_pin_data;
+	int rc = 0;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/* Read Input Register */
+	uint8_t cmd = TCA9538_INPUT_PORT;
+
+	rc = i2c_write_read(cfg->i2c_dev, cfg->i2c_addr, &cmd,
+			    sizeof(cmd), &input_pin_data,
+			    sizeof(input_pin_data));
+	LOG_DBG("read %x got %d", input_pin_data, rc);
+
+	if (rc == 0) {
+		drv_data->pin_state.input = input_pin_data;
+		*value = (gpio_port_value_t)(drv_data->pin_state.input);
+	}
+
+	k_sem_give(&drv_data->lock);
+	return rc;
+}
+
+static int gpio_tca9538_port_write(const struct device *dev,
+				   gpio_port_pins_t mask,
+				   gpio_port_value_t value,
+				   gpio_port_value_t toggle)
+{
+	const struct tca9538_config *cfg = dev->config;
+	struct tca9538_drv_data *drv_data = dev->data;
+	uint8_t *outp = &drv_data->pin_state.output;
+	int rc;
+	char tx_dat[2];
+	uint8_t orig_out;
+	uint8_t out;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	orig_out = *outp;
+	out = ((orig_out & ~mask) | (value & mask)) ^ toggle;
+
+	tx_dat[0] = TCA9538_OUTPUT_PORT;
+	tx_dat[1] = out;
+	rc = i2c_write(cfg->i2c_dev, tx_dat, sizeof(tx_dat),
```

Comment:
`drivers/i2c.h` provides `i2c_reg_write_byte` helper function. The code would be easier to read if we used it. Here and in a few other places. There is also `i2c_reg_read_byte`.